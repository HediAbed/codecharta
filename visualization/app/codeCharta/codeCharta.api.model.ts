import {
	AttributeTypes,
	AttributeTypeValue,
	CodeMapNode,
	Edge,
	MarkedPackage,
	MetricStatisticsItem,
	RecursivePartial,
	Settings
} from "./codeCharta.model"

export interface ExportCCFile {
	projectName: string
	apiVersion: string
	fileChecksum: string
	nodes: CodeMapNode[]
	attributeTypes?: AttributeTypes | OldAttributeTypes
	edges?: Edge[]
	markedPackages?: MarkedPackage[]
	blacklist?: ExportBlacklistItem[]
	metricStatistics?: ExportMetricStatisticsItem[]
}

export interface ExportBlacklistItem {
	path: string
	type: ExportBlacklistType
}

export type ExportMetricStatisticsItem = MetricStatisticsItem

export enum ExportBlacklistType {
	hide = "hide",
	exclude = "exclude"
}

export enum APIVersions {
	ZERO_POINT_ONE = "0.1",
	ONE_POINT_ZERO = "1.0",
	ONE_POINT_ONE = "1.1",
	ONE_POINT_TWO = "1.2"
}

export interface ExportScenario {
	name: string
	settings: RecursivePartial<Settings>
}

export interface OldAttributeTypes {
	nodes?: [{ [key: string]: AttributeTypeValue }?]
	edges?: [{ [key: string]: AttributeTypeValue }?]
}
