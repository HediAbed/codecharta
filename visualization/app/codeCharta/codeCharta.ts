"use strict";

import "./core/core.module";
import "./ui/ui";

import {codeChartaComponent} from "./codeCharta.component";

import angular from "angular";
import { CodeChartaService } from "./codeCharta.service";
import _ from "lodash";

angular.module(
    "app.codeCharta",
    ["app.codeCharta.core", "app.codeCharta.ui"]
);

angular.module("app.codeCharta").component(
    codeChartaComponent.selector,
    codeChartaComponent
).service(
    _.camelCase(CodeChartaService.name),
    CodeChartaService
);


