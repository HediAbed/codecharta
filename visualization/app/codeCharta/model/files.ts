import { CCFile, FileSelectionState, FileState } from "./codeCharta.model"

export class Files {
	private files: FileState[] = []

	public getFiles(): FileState[] {
		return this.files
	}

	public reset() {
		this.files = []
	}

	public resetSelection() {
		this.files.map(file => {
			file.selectedAs = FileSelectionState.None
		})
	}

	public getCCFiles(): CCFile[] {
		return this.files.map(x => x.file)
	}

	public addFile(file: CCFile) {
		this.files.push({ file: file, selectedAs: FileSelectionState.None })
	}

	public setSingleByName(fileName: string) {
		const singleFile: CCFile = this.getFileByFileName(fileName)
		this.setSingle(singleFile)
	}

	public setSingle(file: CCFile) {
		this.resetSelection()
		const matchedFile = this.files.find(x => x.file === file)
		if (matchedFile) {
			matchedFile.selectedAs = FileSelectionState.Single
		}
	}

	public setDeltaByNames(referenceFileName: string, comparisonFileName: string) {
		const referenceFile: CCFile = this.getFileByFileName(referenceFileName)
		const comparisonFile: CCFile = this.getFileByFileName(comparisonFileName)
		this.setDelta(referenceFile, comparisonFile)
	}

	public setDelta(reference: CCFile, comparison: CCFile) {
		this.resetSelection()
		const matchedReferenceFile = this.files.find(x => x.file === reference)
		const matchedComparisonFile = this.files.find(x => x.file === comparison)

		if (matchedReferenceFile) {
			matchedReferenceFile.selectedAs = FileSelectionState.Reference
		}

		if (matchedComparisonFile) {
			matchedComparisonFile.selectedAs = FileSelectionState.Comparison
		}
	}

	public setMultipleByNames(partialFileNames: string[]) {
		const partialFiles: CCFile[] = []
		partialFileNames.forEach(fileName => partialFiles.push(this.getFileByFileName(fileName)))
		this.setMultiple(partialFiles)
	}

	public setMultiple(multipleFiles: CCFile[]) {
		this.resetSelection()
		this.files.filter(x => multipleFiles.indexOf(x.file) !== -1).forEach(x => (x.selectedAs = FileSelectionState.Partial))
	}

	public fileStatesAvailable(): boolean {
		return this.files.length > 0
	}

	public getFileByFileName(fileName: string): CCFile {
		const matchingFileState: FileState = this.files.find(x => x.file.fileMeta.fileName == fileName)
		return matchingFileState ? matchingFileState.file : undefined
	}

	public getVisibleFiles(): CCFile[] {
		return this.files.filter(x => x.selectedAs !== FileSelectionState.None).map(x => x.file)
	}

	public getVisibleFileStates(): FileState[] {
		return this.files.filter(x => x.selectedAs !== FileSelectionState.None)
	}

	public isSingleState(): boolean {
		return !!this.files.find(x => x.selectedAs === FileSelectionState.Single)
	}

	public isDeltaState(): boolean {
		return !!this.files.find(x => x.selectedAs === FileSelectionState.Reference || x.selectedAs === FileSelectionState.Comparison)
	}

	public isPartialState(): boolean {
		return !!this.files.find(x => x.selectedAs === FileSelectionState.Partial)
	}
}
