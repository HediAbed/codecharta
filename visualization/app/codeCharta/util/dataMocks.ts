import { CodeMapNode, Edge, CCFile, Settings, FileSelectionState } from "../codeCharta.model"

export const VALID_NODE: CodeMapNode = {
	name: "root",
	attributes: {},
	type: "Folder",
	children: [
		{
			name: "big leaf",
			type: "File",
			attributes: { RLOC: 100, Functions: 10, MCC: 1 },
			link: "http://www.google.de"
		},
		{
			name: "Parent Leaf",
			type: "Folder",
			attributes: {},
			children: [
				{
					name: "small leaf",
					type: "File",
					attributes: { RLOC: 30, Functions: 100, MCC: 100 }
				},
				{
					name: "other small leaf",
					type: "File",
					attributes: { RLOC: 70, Functions: 1000, MCC: 10 }
				}
			]
		}
	]
}

export const VALID_NODE_WITH_PATH: CodeMapNode = {
	name: "root",
	attributes: {},
	type: "Folder",
	path: "/root",
	children: [
		{
			name: "big leaf",
			type: "File",
			path: "/root/big leaf",
			attributes: { RLOC: 100, Functions: 10, MCC: 1 },
			link: "http://www.google.de"
		},
		{
			name: "Parent Leaf",
			type: "Folder",
			attributes: {},
			path: "/root/Parent Leaf",
			children: [
				{
					name: "small leaf",
					type: "File",
					path: "/root/Parent Leaf/small leaf",
					attributes: { RLOC: 30, Functions: 100, MCC: 100 }
				},
				{
					name: "other small leaf",
					type: "File",
					path: "/root/Parent Leaf/other small leaf",
					attributes: { RLOC: 70, Functions: 1000, MCC: 10 }
				},
				{
					name: "empty folder",
					type: "Folder",
					path: "/root/Parent Leaf/empty folder",
					attributes: {},
					children: []
				}
			]
		}
	]
}

export const VALID_EDGES: Edge[] = [
	{
		fromNodeName: "/root/big leaf",
		toNodeName: "/root/Parent Leaf/small leaf",
		attributes: {
			pairingRate: 89,
			avgCommits: 34
		}
	},
	{
		fromNodeName: "/root/sample1 only leaf",
		toNodeName: "/root/Parent Leaf/small leaf",
		attributes: {
			pairingRate: 89,
			avgCommits: 34
		}
	}
]

export const VALID_EDGE: Edge = {
	fromNodeName: "/root/big leaf",
	toNodeName: "/root/Parent Leaf/small leaf",
	attributes: {
		pairingRate: 89,
		avgCommits: 34
	}
}

export const TEST_FILE_CONTENT = {
	fileName: "noFileName",
	projectName: "Sample Map",
	apiVersion: "1.1",
	nodes: [VALID_NODE]
}

export const TEST_FILE_DATA: CCFile = {
	fileMeta: {
		fileName: "file",
		projectName: "Sample Map",
		apiVersion: "1.1"
	},
	map: VALID_NODE,
	settings: {
		fileSettings: {
			edges: VALID_EDGES
		}
	}
}

export const TEST_FILE_WITH_PATHS: CCFile = {
	fileMeta: {
		fileName: "fileA",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		path: "/root",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				path: "/root/big leaf",
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				path: "/root/Parent Leaf",
				children: [
					{
						name: "small leaf",
						type: "File",
						path: "/root/Parent Leaf/small leaf",
						attributes: { rloc: 30, functions: 100, mcc: 100 }
					},
					{
						name: "other small leaf",
						type: "File",
						path: "/root/Parent Leaf/other small leaf",
						attributes: { rloc: 70, functions: 1000, mcc: 10 }
					},
					{
						name: "empty folder",
						type: "Folder",
						path: "/root/Parent Leaf/empty folder",
						attributes: {},
						children: []
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {}
	}
}

export const TEST_DELTA_MAP_A: CCFile = {
	fileMeta: {
		fileName: "fileA",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				children: [
					{
						name: "small leaf",
						type: "File",
						attributes: { rloc: 30, functions: 100, mcc: 100 }
					},
					{
						name: "other small leaf",
						type: "File",
						attributes: { rloc: 70, functions: 1000, mcc: 10 }
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {}
	}
}

export const TEST_DELTA_MAP_B: CCFile = {
	fileMeta: {
		fileName: "fileB",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				attributes: { rloc: 20, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "additional leaf",
				type: "File",
				attributes: { rloc: 10, functions: 11, mcc: 5 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				children: [
					{
						name: "small leaf",
						type: "File",
						attributes: { rloc: 30, functions: 100, mcc: 100, more: 20 }
					},
					{
						name: "other small leaf",
						type: "File",
						attributes: { rloc: 70, functions: 1000 }
					},
					{
						name: "big leaf",
						type: "File",
						attributes: { rloc: 20, functions: 10, mcc: 1 },
						link: "http://www.google.de"
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {}
	}
}

export const TEST_FILE_DATA_DOWNLOADED = {
	apiVersion: "1.1",
	attributeTypes: undefined,
	blacklist: undefined,
	edges: [
		{ attributes: { avgCommits: 34, pairingRate: 89 }, fromNodeName: "/root/big leaf", toNodeName: "/root/Parent Leaf/small leaf" },
		{
			attributes: { avgCommits: 34, pairingRate: 89 },
			fromNodeName: "/root/sample1 only leaf",
			toNodeName: "/root/Parent Leaf/small leaf"
		}
	],
	fileName: "file.14_12_2018.json",
	nodes: [
		{
			attributes: {},
			children: [
				{ attributes: { Functions: 10, MCC: 1, RLOC: 100 }, link: "http://www.google.de", name: "big leaf", type: "File" },
				{
					attributes: {},
					children: [
						{ attributes: { Functions: 100, MCC: 100, RLOC: 30 }, name: "small leaf", type: "File" },
						{ attributes: { Functions: 1000, MCC: 10, RLOC: 70 }, name: "other small leaf", type: "File" }
					],
					name: "Parent Leaf",
					type: "Folder"
				}
			],
			name: "root",
			type: "Folder"
		}
	],
	projectName: "Sample Map"
}

export const SETTINGS: Settings = {
	fileSettings: { attributeTypes: {}, blacklist: [], edges: [], markedPackages: [] },
	dynamicSettings: {
		areaMetric: "rloc",
		heightMetric: "mcc",
		colorMetric: "mcc",
		focusedNodePath: null,
		searchedNodePaths: [],
		searchPattern: "",
		margin: 48,
		neutralColorRange: { flipped: false, from: 19, to: 67 }
	},
	appSettings: {
		amountOfTopLabels: 31,
		scaling: { x: 1, y: 1.8, z: 1 },
		camera: { x: 0, y: 300, z: 1000 },
		deltaColorFlipped: false,
		enableEdgeArrows: true,
		hideFlatBuildings: true,
		maximizeDetailPanel: false,
		invertHeight: true,
		dynamicMargin: true,
		isWhiteBackground: false,
		whiteColorBuildings: true,
		mapColors: {
			positive: "#69AE40",
			neutral: "#ddcc00",
			negative: "#820E0E",
			selected: "#EB8319",
			defaultC: "#89ACB4",
			positiveDelta: "#69FF40",
			negativeDelta: "#ff0E0E",
			base: "#666666",
			flat: "#AAAAAA",
			lightGrey: "#DDDDDD",
			angularGreen: "#00BFA5",
			markingColors: ["#FF1D8E", "#1d8eff", "#1DFFFF", "#8eff1d", "#8e1dff", "#FFFF1D"]
		}
	},
	treeMapSettings: { mapSize: 500 }
}
