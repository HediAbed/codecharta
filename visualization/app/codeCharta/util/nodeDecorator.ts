"use strict"
import * as d3 from "d3"
import {CCFile, CodeMapNode, MetricData} from "../codeCharta.model"

export class NodeDecorator {

	public static decorateFile(file: CCFile, metricData: MetricData[]): CCFile {
		let decoratedFile: CCFile = this.deepCopy(file)
		this.decorateMapWithMissingObjects(decoratedFile)
		this.decorateMapWithCompactMiddlePackages(decoratedFile)
		this.decorateLeavesWithMissingMetrics(decoratedFile, metricData)
		this.decorateParentNodesWithSumAttributesOfChildren(decoratedFile, metricData)
		return decoratedFile
	}

	private static deepCopy(file: CCFile): CCFile {
		return JSON.parse(JSON.stringify(file))
	}

	public static preDecorateFile(file: CCFile): CCFile {
		// TODO: predecorate origin as well? so in multiple mode the files keep its original origin attribute
		return this.decorateMapWithPathAttribute(file)
	}

	private static decorateMapWithCompactMiddlePackages(file: CCFile) {
		const isEmptyMiddlePackage = current => {
			return (
				current &&
				current.children &&
				current.children.length === 1 &&
				current.children[0].children &&
				current.children[0].children.length > 0
			)
		}

		const rec = current => {
			if (isEmptyMiddlePackage(current)) {
				let child = current.children[0]
				current.children = child.children
				current.name += "/" + child.name
				current.path += "/" + child.name
				if (child.link) {
					current.link = child.link
				}
				current.attributes = child.attributes
				current.deltas = child.deltas
				rec(current)
			} else if (current && current.children && current.children.length > 1) {
				for (let i = 0; i < current.children.length; i++) {
					rec(current.children[i])
				}
			}
		}

		if (file && file.map) {
			rec(file.map)
		}
	}

	private static decorateMapWithPathAttribute(file: CCFile) {
		if (file && file.map) {
			let root = d3.hierarchy<CodeMapNode>(file.map)
			root.each(node => {
				node.data.path = "/" + root.path(node).map(x => x.data.name).join("/")
			})
		}
		return file
	}

	private static decorateMapWithMissingObjects(file: CCFile) {
		if (file && file.map) {
			let root = d3.hierarchy<CodeMapNode>(file.map)
			root.each(node => {
				node.data.visible = true
				node.data.origin = file.fileMeta.fileName
				node.data.attributes = (!node.data.attributes) ? {} : node.data.attributes
				Object.assign(node.data.attributes, { unary: 1 })
			})
		}
	}

	private static decorateLeavesWithMissingMetrics(file: CCFile, metricData: MetricData[]) {
		if (file && file.map && metricData) {
			let root = d3.hierarchy<CodeMapNode>(file.map)
			root.leaves().forEach(node => {
				metricData.forEach(metric => {
					if (!node.data.attributes.hasOwnProperty(metric.name)) {
						node.data.attributes[metric.name] = 0
					}
				})
			})
		}
	}

	private static decorateParentNodesWithSumAttributesOfChildren(file: CCFile, metricData: MetricData[]) {
		if (file && file.map) {
			let root = d3.hierarchy<CodeMapNode>(file.map)
			root.each(node => {
				this.decorateNodeWithChildrenSumMetrics(node, metricData)
			})
		}
	}

	private static decorateNodeWithChildrenSumMetrics(node, metricData: MetricData[]) {
		metricData.forEach(metric => {
			if (!node.data.attributes.hasOwnProperty(metric.name) && node.data.children && node.data.children.length > 0) {
				this.defineAttributeAsSumMethod(node, metric.name)
			}
		})
	}

	private static defineAttributeAsSumMethod(node, metric: string) {
		Object.defineProperty(node.data.attributes, metric, {
			enumerable: true,
			get: () => {
				return node.leaves()
					.map(x => x.data.attributes[metric])
					.reduce((partialSum, a) => partialSum + a)
			}
		})
	}
}
