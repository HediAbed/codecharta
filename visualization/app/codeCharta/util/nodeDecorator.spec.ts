import { hierarchy, HierarchyNode } from "d3"
import { DEFAULT_STATE, STATE, TEST_DELTA_MAP_A, VALID_NODE_WITH_PATH_AND_DELTAS } from "./dataMocks"
import { AttributeTypeValue, BlacklistItem, BlacklistType, CCFile, CodeMapNode, NodeType, State } from "../codeCharta.model"
import { NodeDecorator } from "./nodeDecorator"
import _ from "lodash"
import { NodeMetricDataService } from "../state/store/metricData/nodeMetricData/nodeMetricData.service"
import { FileSelectionState } from "../model/files/files"
import { clone } from "./clone"

describe("nodeDecorator", () => {
	let file: CCFile
	let map: CodeMapNode
	let deltaMap: CodeMapNode
	let state: State

	beforeEach(() => {
		file = clone(TEST_DELTA_MAP_A)
		map = file.map
		deltaMap = clone(VALID_NODE_WITH_PATH_AND_DELTAS)
		state = clone(DEFAULT_STATE)
		state.metricData.nodeMetricData = [
			{ name: "rloc", maxValue: 999999 },
			{ name: "functions", maxValue: 999999 },
			{
				name: "mcc",
				maxValue: 999999
			}
		]
		state.metricData.edgeMetricData = [
			{ name: "pairingRate", maxValue: 999 },
			{ name: "avgCommits", maxValue: 999 }
		]
		state.fileSettings.attributeTypes = {
			nodes: { functions: AttributeTypeValue.relative, rloc: AttributeTypeValue.absolute },
			edges: { pairingRate: AttributeTypeValue.relative }
		}
		state.fileSettings.blacklist = clone(STATE.fileSettings.blacklist)
		state.files = [{ selectedAs: FileSelectionState.Single, file }]
		NodeDecorator.decorateMapWithPathAttribute(file)
	})

	function allUniqueIds(map: HierarchyNode<CodeMapNode>): boolean {
		const ids = new Set()
		map.each(node => {
			if (ids.has(node.id)) {
				return false
			}
			ids.add(node.id)
		})
		return true
	}

	describe("decorateMap", () => {
		it("leaves should have all metrics", () => {
			state.metricData.nodeMetricData.push({ name: "some", maxValue: 999999 })
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)

			hierarchy(map)
				.leaves()
				.forEach(node => {
					expect(node.data.attributes).toBeDefined()
					expect(node.data.attributes.some).toBe(0)
					expect(node.data.attributes.rloc).toBeDefined()
					expect(node.data.attributes.functions).toBeDefined()
					expect(node.data.attributes.mcc).toBeDefined()
				})
		})

		it("leaves should have all metrics even if some attributesLists are undefined", () => {
			map.children[0].attributes = undefined
			state.metricData.nodeMetricData.push({ name: "some", maxValue: 999999 })

			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)
			hierarchy(map)
				.leaves()
				.forEach(node => {
					expect(node.data.attributes).toBeDefined()
					expect(node.data.attributes.some).toBe(0)
					expect(node.data.attributes.rloc).toBeDefined()
					expect(node.data.attributes.functions).toBeDefined()
					expect(node.data.attributes.mcc).toBeDefined()
				})
		})

		it("should compact from root", () => {
			map.children = [
				{
					name: "middle",
					type: NodeType.FOLDER,
					attributes: {},
					isExcluded: false,
					isFlattened: false,
					children: [
						{
							name: "a",
							type: NodeType.FILE,
							attributes: {},
							isExcluded: false,
							isFlattened: false
						},
						{
							name: "b",
							type: NodeType.FILE,
							attributes: {},
							isExcluded: false,
							isFlattened: false
						}
					]
				}
			]
			NodeDecorator.decorateMapWithPathAttribute(file)
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)
			expect(map.name).toBe("root/middle")
			expect(map.children.length).toBe(2)
			expect(map.children[0].name).toBe("a")
			expect(map.children[1].name).toBe("b")
		})

		it("should collect links correctly", () => {
			map.link = "link0"
			map.children = [
				{
					name: "middle",
					type: NodeType.FILE,
					attributes: {},
					link: "link1",
					isExcluded: false,
					isFlattened: false,
					children: [
						{
							name: "a",
							type: NodeType.FILE,
							attributes: {},
							isExcluded: false,
							isFlattened: false
						},
						{
							name: "b",
							type: NodeType.FILE,
							attributes: {},
							isExcluded: false,
							isFlattened: false
						}
					]
				}
			]
			NodeDecorator.decorateMapWithPathAttribute(file)
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)
			expect(map.link).toBe("link1")
		})

		it("should collect paths correctly", () => {
			map.path = "/root"
			map.children = [
				{
					name: "middle",
					path: "/root/middle",
					type: NodeType.FOLDER,
					attributes: {},
					isExcluded: false,
					isFlattened: false,
					children: [
						{
							name: "a",
							type: NodeType.FILE,
							path: "/root/middle/a",
							attributes: {},
							isExcluded: false,
							isFlattened: false
						},
						{
							name: "b",
							type: NodeType.FILE,
							path: "/root/middle/b",
							attributes: {},
							isExcluded: false,
							isFlattened: false
						}
					]
				}
			]
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)
			expect(map.path).toBe("/root/middle")
		})

		it("should not compact with single leaves", () => {
			map.children = [
				{
					name: "middle",
					type: NodeType.FOLDER,
					attributes: {},
					isExcluded: false,
					isFlattened: false,
					children: [
						{
							name: "singleLeaf",
							type: NodeType.FILE,
							attributes: {},
							isExcluded: false,
							isFlattened: false
						}
					]
				}
			]
			NodeDecorator.decorateMapWithPathAttribute(file)
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)
			expect(map.name).toBe("root/middle")
			expect(map.children.length).toBe(1)
			expect(map.children[0].name).toBe("singleLeaf")
		})

		it("should compact intermediate middle packages", () => {
			map.children = [
				{
					name: "start",
					type: NodeType.FOLDER,
					attributes: {},
					isExcluded: false,
					isFlattened: false,
					children: [
						{
							name: "middle",
							type: NodeType.FOLDER,
							attributes: {},
							isExcluded: false,
							isFlattened: false,
							children: [
								{
									name: "middle2",
									type: NodeType.FOLDER,
									attributes: {},
									isExcluded: false,
									isFlattened: false,
									children: [
										{
											name: "a",
											type: NodeType.FILE,
											attributes: {},
											isExcluded: false,
											isFlattened: false
										},
										{
											name: "b",
											type: NodeType.FILE,
											attributes: {},
											isExcluded: false,
											isFlattened: false
										}
									]
								}
							]
						},
						{
							name: "c",
							type: NodeType.FILE,
							attributes: {},
							isExcluded: false,
							isFlattened: false
						}
					]
				}
			]
			NodeDecorator.decorateMapWithPathAttribute(file)
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)
			expect(map.name).toBe("root/start")
			expect(map.children.length).toBe(2)
			expect(map.children[0].name).toBe("middle/middle2")
			expect(map.children[1].name).toBe("c")
			expect(map.children[0].children.length).toBe(2)
			expect(map.children[0].children[0].name).toBe("a")
			expect(map.children[0].children[1].name).toBe("b")
		})

		it("maps with no attribute nodes should be accepted and an attributes member added", () => {
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)

			hierarchy(map).each(node => {
				expect(node.data.attributes[NodeMetricDataService.UNARY_METRIC]).toBeDefined()
			})
		})

		it("all nodes should have a unary attribute", () => {
			map.children[0].attributes = {}
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)

			hierarchy(map).each(node => {
				expect(node.data.attributes[NodeMetricDataService.UNARY_METRIC]).toBeDefined()
			})
		})

		it("should decorate nodes with a unique id starting from 0", () => {
			NodeDecorator.decorateMap(file.map, [])

			const h = hierarchy(file.map)
			hierarchy(map).each(node => {
				expect(node.data.id).toBeDefined()
				expect(allUniqueIds(h)).toBeTruthy()
			})
			expect(file.map.id).toBe(0)
		})
	})

	describe("decorateMapWithPathAttribute", () => {
		it("should decorate nodes with the correct path", () => {
			NodeDecorator.decorateMapWithPathAttribute(file)

			hierarchy(file.map).each(node => {
				expect(node.data.path).toBeDefined()
			})

			expect(file.map.path).toBe("/root")
			expect(file.map.children[1].children[0].path).toBe("/root/Parent Leaf/small leaf")
		})
	})

	describe("decorateParentNodesWithAggregatedAttributes", () => {
		it("should aggregate given absolute metrics correctly", () => {
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)
			NodeDecorator.decorateParentNodesWithAggregatedAttributes(map, state)

			expect(map.attributes["rloc"]).toBe(200)
			expect(map.attributes["mcc"]).toBe(111)
		})

		it("should aggregate given relative metrics correctly", () => {
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)
			NodeDecorator.decorateParentNodesWithAggregatedAttributes(map, state)

			expect(map.attributes["functions"]).toBe(100)
		})

		it("should aggregate absolute edge metrics correctly", () => {
			map.children[0].edgeAttributes = { avgCommits: { incoming: 12, outgoing: 13 } }
			map.children[1].children[0].edgeAttributes = { avgCommits: { incoming: 10, outgoing: 10 } }
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)

			NodeDecorator.decorateParentNodesWithAggregatedAttributes(map, state)

			expect(map.edgeAttributes["avgCommits"].incoming).toBe(22)
			expect(map.edgeAttributes["avgCommits"].outgoing).toBe(23)
		})

		it("should aggregate given relative edge metrics correctly", () => {
			map.children[0].edgeAttributes = { pairingRate: { incoming: 12, outgoing: 13 } }
			map.children[1].children[0].edgeAttributes = { pairingRate: { incoming: 10, outgoing: 10 } }
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)

			NodeDecorator.decorateParentNodesWithAggregatedAttributes(map, state)

			expect(map.edgeAttributes["pairingRate"].incoming).toBe(11)
			expect(map.edgeAttributes["pairingRate"].outgoing).toBe(11.5)
		})

		it("should aggregate missing metrics correctly", () => {
			state.metricData.nodeMetricData.push({ name: "some", maxValue: 999999 })
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)
			NodeDecorator.decorateParentNodesWithAggregatedAttributes(map, state)

			expect(map.attributes["rloc"]).toBe(200)
			expect(map.attributes["some"]).toBe(0)
			expect(map.attributes["some other attribute"]).not.toBeDefined()
		})

		it("all nodes should have an attribute list with all possible metrics", () => {
			map.children[0].attributes = undefined
			map.children[1].attributes = { some: 1 }
			state.metricData.nodeMetricData.push({ name: "some", maxValue: 999999 })

			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)
			NodeDecorator.decorateParentNodesWithAggregatedAttributes(map, state)

			hierarchy(map).each(node => {
				expect(node.data.attributes).toBeDefined()
				expect(node.data.attributes.some).toBeDefined()
			})
		})

		it("all nodes should have an attribute list with listed and available metrics", () => {
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)
			NodeDecorator.decorateParentNodesWithAggregatedAttributes(map, state)

			hierarchy(map).each(node => {
				expect(node.data.attributes).toBeDefined()
				expect(node.data.attributes["rloc"]).toBeDefined()
				expect(node.data.attributes["functions"]).toBeDefined()
			})
		})

		it("folders should have sum attributes of children for absolute metrics", () => {
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)
			NodeDecorator.decorateParentNodesWithAggregatedAttributes(map, state)

			const h = hierarchy(map)

			expect(h.data.attributes["rloc"]).toBe(200)
			expect(h.children[0].data.attributes["rloc"]).toBe(100)
		})

		it("folders should have median attributes of children for relative metrics", () => {
			NodeDecorator.decorateMap(map, state.metricData.nodeMetricData)
			NodeDecorator.decorateParentNodesWithAggregatedAttributes(map, state)

			expect(hierarchy(map).data.attributes["functions"]).toBe(100)
		})

		it("folders should have sum delta values of children for absolute metrics", () => {
			state.files[0].selectedAs = FileSelectionState.Comparison
			NodeDecorator.decorateMap(deltaMap, state.metricData.nodeMetricData)
			NodeDecorator.decorateParentNodesWithAggregatedAttributes(deltaMap, state)

			const h = hierarchy(deltaMap)

			expect(h.data.deltas["rloc"]).toBe(295)
			expect(h.children[0].data.deltas["rloc"]).toBe(300)
			expect(h.children[2].data.deltas["rloc"]).toBe(145)
		})

		it("folders should have median delta values of children for relative metrics", () => {
			state.files[0].selectedAs = FileSelectionState.Comparison

			NodeDecorator.decorateMap(deltaMap, state.metricData.nodeMetricData)
			NodeDecorator.decorateParentNodesWithAggregatedAttributes(deltaMap, state)

			expect(hierarchy(deltaMap).data.deltas["functions"]).toBe(-3)
		})
	})

	describe("decorateMapWithBlacklist", () => {
		it("should update all flattened and excluded attributes on the map based on the blacklist", () => {
			state.fileSettings.blacklist.push({ path: "**/Parent Leaf", type: BlacklistType.flatten })

			NodeDecorator.decorateMapWithBlacklist(map, state.fileSettings.blacklist)

			expect(map).toMatchSnapshot()
		})

		it("should revert the map back to it's previous state, if all buildings will be excluded", () => {
			const item: BlacklistItem = { path: "*", type: BlacklistType.exclude }
			const oldMap = _.cloneDeep(map)

			try {
				NodeDecorator.decorateMapWithBlacklist(map, [item])
			} catch (e) {
				expect(e.message).toEqual("Excluding all buildings is not possible.")
				expect(map).toEqual(oldMap)
			}
		})
	})
})
