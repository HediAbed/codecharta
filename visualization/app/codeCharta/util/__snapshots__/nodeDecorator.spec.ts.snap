// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodeDecorator decorateMapWithBlacklist should update all flattened and excluded attributes on the map based on the blacklist 1`] = `
Object {
  "attributes": Object {},
  "children": Array [
    Object {
      "attributes": Object {
        "functions": 10,
        "mcc": 1,
        "rloc": 100,
      },
      "isExcluded": false,
      "isFlattened": false,
      "link": "http://www.google.de",
      "name": "big leaf",
      "path": "/root/big leaf",
      "type": "File",
    },
    Object {
      "attributes": Object {},
      "children": Array [
        Object {
          "attributes": Object {
            "functions": 100,
            "mcc": 100,
            "rloc": 30,
          },
          "isExcluded": false,
          "isFlattened": true,
          "name": "small leaf",
          "path": "/root/Parent Leaf/small leaf",
          "type": "File",
        },
        Object {
          "attributes": Object {
            "functions": 1000,
            "mcc": 10,
            "rloc": 70,
          },
          "isExcluded": false,
          "isFlattened": true,
          "name": "other small leaf",
          "path": "/root/Parent Leaf/other small leaf",
          "type": "File",
        },
      ],
      "isExcluded": false,
      "isFlattened": true,
      "name": "Parent Leaf",
      "path": "/root/Parent Leaf",
      "type": "Folder",
    },
  ],
  "isExcluded": false,
  "isFlattened": false,
  "name": "root",
  "path": "/root",
  "type": "Folder",
}
`;
